### PR TITLE
Allow device overrides when initialising or launching an instance.

### LINKED ISSUE #10796: Add support for device overrides when creating/launching instance
Suggested at https://discuss.linuxcontainers.org/t/how-to-make-ips-fixed-to-an-instance-from-the-managed-bridge/14838/4?u=tomp

Proposal to add `--device` and `-d` flags to `lxc launch` and `lxc init` that would allow device override at create time.
Similar to what we have now for `lxc copy` and `lxc move`.

The format of the flag value would be the same, i.e `-d <device_name>,<key>=<value>`.

### PR DESCRIPTION
**Pull request by markylaing**
Adds a `--device`/`-d` flag to `lxc init` and `lxc launch` to pass in device overrides of the form `<deviceName>,<key>=<value>`. The overrides are performed after the instance is created, if an overridden device is not found in the expanded devices of the instance (i.e. devices applied via profile), then the instance is deleted.

Closes #10796

### CODE REVIEW

File: lxc/init.go [hunk 1]
```diff
@@ -370,6 +381,39 @@ func (c *cmdInit) create(conf *config.Config, args []string) (lxd.InstanceServer
 	// Validate the network setup
 	c.checkNetwork(d, name)
 
+	if deviceOverrides != nil {
```
Comment (tomponline) [inline lxc/init.go:384]: hey @markylaing  I've not had chance yet to look at this in detail, but this doesn't look like the correct approach.
If I'm reading this correctly, you're updating the instance's config after it has been created.
I think we need to override the config and then include it in the instance creation request itself.

Comment (markylaing) [reply lxc/init.go:384]: I'm confused. This is what I was asking here https://github.com/lxc/lxd/issues/10796#issuecomment-1288924934

I can't get the expanded devices until the instance has been created... I'd need to check which profiles are going to be applied in advance.

Comment (tomponline) [reply lxc/init.go:384]: Yeah I can see why you got confused.

What we meant is that is effectively what needs to be done, but I need to look at the API to see if its possible for us to "effectively" do at create time.

Comment (markylaing) [reply lxc/init.go:384]: Ok. We need to query for the profiles that are to be applied in advance, then override any matching devices passed via the `--device` flag. (That was option 1 in https://github.com/lxc/lxd/issues/10796#issuecomment-1288924934)

Comment (tomponline) [reply lxc/init.go:384]: yeah potentially we need to "expand" the provisional device config and then copy and apply changes to the specific device(s) to send as the device config for the new instance.

Comment (markylaing) [reply lxc/init.go]: Ok I think I've made the changes.

**Comment by markylaing** [comment]
@stgraber @tomponline while working on this PR I've found the behaviour of the `--no-profiles` flag quite confusing. The default profile seems to be applied regardless of whether the flag is set or not. Is this correct? If so, what is the difference between adding `--no-profiles` and simply not specifying any profiles with `--profile`?

**Comment by tomponline** [comment]
Good question! I'd have to look at the code and help text, although @stgraber may have additional context as to the original intention of that flag.

**Review by tomponline — CHANGES_REQUESTED** [review]
This looks on the right track now, thanks.

Although tests are failing, and please can you add a test for this new functionality?

Thanks

**Comment by markylaing** [comment]
> This looks on the right track now, thanks.
> 
> Although tests are failing, and please can you add a test for this new functionality?
> 
> Thanks

Tests should be fixed now and I've added a simple test which overrides the storage pool of the root disk device on `init`. (This isn't possible via `lxc config device override root pool=<newPool>`).

File: test/suites/basic.sh [hunk 2]
```diff
@@ -547,7 +548,17 @@ test_basic_usage() {
   ! lxc profile rename default foobar || false
   ! lxc profile delete default || false
 
+  # Should fail to override root device storage pool when container is already initialised.
```
Comment (tomponline) [inline test/suites/basic.sh:551]: This should be:

> # Should fail to override device if it already exists in the instance.

As oppose to the container already being initialised.

File: test/suites/basic.sh [hunk 3]
```diff
@@ -547,7 +548,17 @@ test_basic_usage() {
...
   lxc init testimage c1
   ! lxc config device override c1 root pool bla || false
   lxc rm -f c1
+
+  # Should fail to override root device storage pool when the new pool does not exist.
+  ! lxc init testimage c1 -d root,pool=bla || false
+
+  # Should succeed in overriding root device storage pool when the pool does exist and the override occurs at create time.
+  lxc storage create bla dir
+  lxc init testimage c1 -d root,pool=bla
```
Comment (tomponline) [inline test/suites/basic.sh:561]: Can you add a check for `lxc config show <instance> --expanded` here to check the pool override took effect in the expanded config?

Comment (markylaing) [reply test/suites/basic.sh:420]: Done.

File: test/suites/basic.sh (hunk 2 above)
Comment (markylaing) [reply test/suites/basic.sh]: I've added test cases for both the situations we discussed and grepped for the specific error we expect.

**Review by tomponline — APPROVED** [review]

**Pull request merged**